Add write support for other file formats in HiveConnector

majetideepak opened the pull request:
HiveConnector today only supports the DWRF file format. 
This PR generalizes it to other file formats such as Parquet.
Add `dwio::common::Writer` API and extend it to DWRF and Parquet file formats.
Add support for Writer Factories.
Extend HiveDataSink to utilize writer factories.

The `WriterOptions` for DWRF and parquet have nothing in common.
The `dwrf::WriterOptions` is also strongly coupled with DWRF structures.
`dwio::common::WriterOptions` API has been created to store some common fields.
The current implementation specializes `WriterOptions` for DWRF and Parquet.

Resolves: https://github.com/facebookincubator/velox/issues/4874

majetideepak commented:
@Yuhta, @gggrace14 Can you please help review/merge this?

Yuhta commented:
@xiaoxmeng Can you help review and merge this one?

xiaoxmeng commented:
> @xiaoxmeng Can you help review and merge this one?

@Yuhta @majetideepak sure, will take care!

xiaoxmeng commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -188,21 +183,21 @@ void HiveDataSink::ensurePartitionWriters() {
+  dwio::common::WriterOptions options;
+  options.schema = inputType_;
+  options.memoryPool = connectorQueryCtx_->connectorMemoryPool();
+  writers_.push_back(writerFactory->createWriter(std::move(sink), options));
```
nit: writers_.push_back(writerFactory->createWriter(dwio::common::DataSink::create(writePath)),

xiaoxmeng commented on velox/connectors/hive/HiveDataSink.cpp:
```diff
@@ -188,21 +183,21 @@ void HiveDataSink::ensurePartitionWriters() {
+  auto sink = dwio::common::DataSink::create(writePath);
+
+  auto writerFactory =
+      dwio::common::getWriterFactory(insertTableHandle_->tableStorageFormat());
```
```
s/tableStorageFormat/fileFormat/?
```

majetideepak replied:
I believe the protocol has a provision for partitionStorageFormat also. So `tableStorageFormat` simply follows the protocol.
https://github.com/prestodb/presto/blob/e825806e6b02e601cc635100514a9442581c05ed/presto-native-execution/presto_cpp/presto_protocol/presto_protocol.h#L2645

xiaoxmeng commented on velox/connectors/hive/HiveDataSink.h:
```diff
@@ -84,13 +88,18 @@ class HiveInsertTableHandle : public ConnectorInsertTableHandle {
     return locationHandle_;
   }
 
+  const dwio::common::FileFormat tableStorageFormat() const {
```
```
nit: s/tableStorageFormat/fileFormat/
```

majetideepak replied:
same as https://github.com/facebookincubator/velox/pull/5055#discussion_r1218317621

xiaoxmeng commented on velox/dwio/common/WriterFactory.cpp:
```diff
@@ -0,0 +1,61 @@
+} // namespace
+
+bool registerWriterFactory(std::shared_ptr<WriterFactory> factory) {
+  bool ok = writerFactories().insert({factory->fileFormat(), factory}).second;
```
```
nit: const bool ok =
```

xiaoxmeng commented on velox/dwio/common/WriterFactory.cpp:
```diff
@@ -0,0 +1,61 @@
+}
+
+bool unregisterWriterFactory(FileFormat format) {
+  auto count = writerFactories().erase(format);
```
```
return writerFactories().erase(format) == 1;
```

xiaoxmeng commented on velox/dwio/common/WriterFactory.cpp:
```diff
@@ -0,0 +1,61 @@
+
+bool hasWriterFactory(FileFormat format) {
+  auto it = writerFactories().find(format);
+  return it != writerFactories().end();
```
```
return writerFactories().count(format) == 1;
```

xiaoxmeng reviewed (approved):
@majetideepak Looks pretty good % minors. Thanks for the refactoring!

majetideepak commented on velox/dwio/parquet/tests/duckdb_reader/ParquetTableScanTest.cpp:
```diff
@@ -35,6 +35,7 @@ class ParquetTableScanTest : public HiveConnectorTestBase {
 
   void SetUp() override {
     HiveConnectorTestBase::SetUp();
+    unregisterParquetReaderFactory();
```
The base class `HiveConnectorTestBase::Setup()` in this change registers the native parquet reader.
To test the DuckDB Parquet reader, we need to unregister that and register the DuckDB Parquet Reader.
The DuckDB Parquet reader is going away in https://github.com/facebookincubator/velox/issues/5108.
So this will go away as part of that PR.

majetideepak commented on velox/connectors/hive/HiveDataSink.h:
```diff
@@ -69,9 +71,11 @@ class HiveInsertTableHandle : public ConnectorInsertTableHandle {
       std::vector<std::shared_ptr<const HiveColumnHandle>> inputColumns,
-      std::shared_ptr<const LocationHandle> locationHandle)
+      std::shared_ptr<const LocationHandle> locationHandle,
+      const dwio::common::FileFormat tableStorageFormat)
```
I am using the same name as in the PrestoProtocol.
https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/presto_protocol/presto_protocol.h#L2644

majetideepak commented:
@xiaoxmeng thanks again for your review and comments. I addressed some of them and responded to others. I also added a default `dwio::common::FileFormat::DWRF` to the `tableStorageFormat` for `HiveInsertTableHandle` introduced here to make the integration with Prestissimo easier.

majetideepak commented:
> Does dwrf writer needs to inherit WriterBase or not?

@xiaoxmeng I saw a similar change on the reader side. So I felt this was safe.
https://github.com/facebookincubator/velox/commit/71805dfed5248bd0c9e5ed40cf6f6e37e3afa372#diff-8b43ed5b1599de687d2ec7778e66410234c6cc2fe646aaff85352fd6ec4dab5fL109

iamkroot commented:
This seems to cause a build failure with `-DVELOX_ENABLE_PARQUET=ON` flag. 
The `Writer` class defined at [`velox/dwio/parquet/writer/Writer.h`](https://github.com/facebookincubator/velox/blob/7e33788bb11c8be17124da62e4b7ab9ea9ae8b4e/velox/dwio/parquet/writer/Writer.h#L46) overrides the `close` and `flush` functions but they are not marked as such.

```
In file included from velox/dwio/parquet/RegisterParquetWriter.cpp:20:
./velox/dwio/parquet/writer/Writer.h:66:8: error: 'flush' overrides a member function but is not marked 'override' [-Werror,-Winconsistent-missing-override]
  void flush();
       ^
./velox/dwio/common/Writer.h:50:16: note: overridden virtual function is here
  virtual void flush() = 0;
               ^
In file included from velox/dwio/parquet/RegisterParquetWriter.cpp:20:
./velox/dwio/parquet/writer/Writer.h:74:8: error: 'close' overrides a member function but is not marked 'override' [-Werror,-Winconsistent-missing-override]
  void close();
       ^
./velox/dwio/common/Writer.h:56:16: note: overridden virtual function is here
  virtual void close() = 0;
               ^
2 errors generated.
```

majetideepak commented:
Fixed here https://github.com/facebookincubator/velox/pull/5282
Thanks for reporting.
